### PR TITLE
BAK-1193 Backend Litigation (Paused)

### PR DESCRIPTION
**Pull request by diversemix**
# Backend Litigation Work

## Overview
Work status:
 - [x] A first cut of the schema, see `backend/app/db/models/litigation.py` and `backend/app/db/models/case.py`
 - [x] A sneaky change for removing the use of `SmallInteger` - see `backend/app/db/models/document.py` and `backend/app/db/models/geography.py`
 - [x] CSV Exports of the data from [this google sheet](https://docs.google.com/spreadsheets/d/1B_jNrr_yEaETGZAn89LkqIE3XfOA1t69wVSgLCxZ-9c/edit#gid=0) see `backend/scripts/litigation/data`
 - [x] Validation of the cross-referencing of the CSV - see `backend/scripts/litigation/verify.py`
 - [x] Transforming the CSV files into a single JSON object of Cases - see `backend/scripts/litigation/transform.py`
 - [ ] Partially complete - Ingesting the JSON Cases into the database - see `backend/scripts/litigation/transform_and_load.py`
 
The work on the endpoints for bulk importing has not yet started, This needs to mirror the pattern used for law & policies - see [this PR](https://github.com/climatepolicyradar/navigator/pull/805) which added `"/bulk-imports/cclw/law-policy"`
 
## Notable Notes
- @marcusodavies is supporting this work from a data point of view. We need to take a view on:
  - Data with unexpected multiple values : pillar = "Mitigation and Adaptation & Resilience"
  - Values with "&" vs "and"
  - Values with case discrepancies. (should we ignore case)
- @joel-wright and @diversemix still need to 
    - confirm what values are ENUMs (infrequently changed data) and what will need their own tables (more frequently / admin changed)
    - If keywords being and array of text is ok
    - `verify()` just checks the relationship - This should also check the values prior to `transform()`
    - Need to verify what should cascade in the schema
- Need to write tests 

## Setup for Development
Aliases used
```
alias reset='make stop; docker-compose down; docker volume rm navigator_db-data-backend; make start'
alias alembic_gen='alembic revision --autogenerate -m'
alias root='cd $(git rev-parse --show-toplevel)'
```

### Regenerating the Schema
```
root
rm backend/alembic/versions/0006_litigation.py
reset
cd backend
PYTHONPATH=$PWD alembic_gen litigation
```

### Running the scripts (with [`entr`](https://github.com/clibs/entr))
```
root
cd backend
find $PWD| PYTHONPATH=$PWD entr python scripts/litigation/transform_and_load.py
```

### CODE REVIEW

File: backend/app/db/crud/litigation.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    StrategicAlignmentClass,
+    UNFCCCPillars,
+)
+from app.db.models.geography import Geography
+from app.db.models.document import Sector
+
+from sqlalchemy.orm import Session
+
+
+def _to_bool(value: str) -> bool:
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:17]: Probably should be in some utility module somewhere

File: backend/app/db/crud/litigation.py [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    entity_name: str,
+    entity_key: str,
+    primary_set: Set[str],
+):
+    log.info(f"Checking {entity_name}:")
+    referenced_cases = set([e[entity_key] for e in entities.values()])
+    unknown_cases = referenced_cases - primary_set
+    orphaned_cases = primary_set - referenced_cases
+
+    log.info(f"  Found {len(referenced_cases)} referenced {entity_key}s.")
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:33]: Logging is currently tailored for stdout - need to address this

File: backend/app/db/crud/litigation.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        json_cases[cid][LitSideType.RESPONDING_PARTY.value] = []
+        json_cases[cid][LitSideType.INTERVENOR.value] = []
+
+    # Add events
+    for _, e in events.items():
+        cid = e["Case ID"]
+        if cid not in json_cases.keys():
+            msg = f"Event {e} references a non-existent case {cid}"
+            log.error(msg)
+            # raise ValueError(f"Event {e} references a non-existent case {cid}")  - uncomment to abort instead
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:93]: Moving fast at the minute so raising exceptions should be put back before release

File: backend/app/db/crud/litigation.py [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        party_type=party_type,
+        side_type=side_type,
+    )
+
+    db.add(new_party)
+    db.commit()
+    return new_party
+
+
+def ingest_case(db: Session, log: logging.Logger, json_case: Mapping[str, Any]) -> Case:
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:158]: Still need to sort out the db transactions - see what Joel did :smile:

File: backend/app/db/models/case.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    CLOSED_SETTLED = "Closed - settled"
+    CLOSED_WITHDRAWN = "Closed - withdrawn"
+    CLOSED = "Closed"
+
+    @staticmethod
+    def from_value(value: str) -> Union["CaseStatus", None]:
+        """Creates the enum from a string value."""
+        if value is None:
+            return None
+        if value == "N/A":
```
Comment (diversemix) [inline backend/app/db/models/case.py:42]: Confirm with @marcusodavies

File: backend/app/db/models/case.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    LOSS_DAMAGE = "Loss & Damage"
+
+    @staticmethod
+    def from_value(value: str) -> Union["UNFCCCPillars", None]:
+        """Creates the enum from a string value."""
+        if value is None:
+            return None
+        if value.upper() == "LOSS AND DAMAGE":
+            return UNFCCCPillars.LOSS_DAMAGE
+        if value == "Mitigation and Adaptation & Resilience":
```
Comment (diversemix) [inline backend/app/db/models/case.py:61]: Confirm with @marcusodavies

File: backend/app/db/models/case.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    objective = Column(Text)
+    pillars = Column(Enum(UNFCCCPillars))
+    summary = Column(Text)
+    reference = Column(Text)
+    case_class = Column(Text)
+    strategic = Column(Boolean)
+    climate_class = Column(Enum(ClimateAlignmentClass))
+    strategic_class = Column(Enum(StrategicAlignmentClass))
+    source = Column(Text)
+    keywords = Column(ARRAY(Text))
```
Comment (diversemix) [inline backend/app/db/models/case.py:141]: Confirm with @joel-wright

File: backend/app/db/models/case.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    __tablename__ = "lit_case"
+
+    id = Column(Integer, primary_key=True)
+    ext_id = Column(Text)
+    name = Column(Text, unique=True, nullable=False)
+    year = Column(Integer)
+    status = Column(Enum(CaseStatus))
+    outcome = Column(Enum(CaseOutcome))
+    objective = Column(Text)
+    pillars = Column(Enum(UNFCCCPillars))
```
Comment (diversemix) [inline backend/app/db/models/case.py:133]: All `Enum`s need confirming with @joel-wright

Comment (diversemix) [inline backend/app/db/models/case.py:133]: Also we can have multiple pillars

File: backend/app/db/crud/litigation.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+    case = Case(
+        name=json_case["Case name"],
+        ext_id="CCLW:" + json_case["Case ID"],
+        year=json_case["Year of filing"],
+        status=CaseStatus.from_value(json_case["Current status"]),
+        outcome=json_case["Assessment of outcome"],
+        objective=json_case["Core objective"],
+        pillars=UNFCCCPillars.from_value(
+            json_case["UNFCCC pillars"].split(";")[0]
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:196]: Schema needs changing

File: backend/app/db/crud/litigation.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        ),  # Schema change?
+        summary=json_case["Summary"],
+        reference=json_case["Citation/reference number"],
+        strategic=_to_bool(json_case["Classified as strategic case"]),
+        climate_class=ClimateAlignmentClass.from_value(climate_class),
+        strategic_class=StrategicAlignmentClass.from_value(
+            strategic_class.split(",")[0]
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:203]: Schema needs changing

File: backend/app/db/crud/litigation.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        .filter(Geography.value == json_case["Country ISO"])
+        .scalar()
+    )
+    if geography_id is None:
+        log.error(
+            f"Geography '{json_case['Country ISO']}' missing for case '{json_case['Case name']}' ({json_case['Case ID']}) was invalid: "
+        )
+
+    # Validate sectors
+    for s in json_case["Sector"].split(";"):
```
Comment (diversemix) [inline backend/app/db/crud/litigation.py:177]: Schema needs changing to support multiple sectors

File: backend/app/db/models/case.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    CLIMATE_ALIGNED = "Climate aligned"
+    CLIMATE_NON_ALIGNED = "Climate non-aligned"
+    JUST_TRANSITION = "Just transition"
+
+    @staticmethod
+    def from_value(value: str) -> Union["ClimateAlignmentClass", None]:
+        """Creates the enum from a string value."""
+        if value is None:
+            return None
+        if value == "N/A":
```
Comment (diversemix) [inline backend/app/db/models/case.py:87]: Confirm with @marcusodavies

File: backend/app/db/models/case.py (hunk 7 above)
Comment (joel-wright) [reply backend/app/db/models/case.py:141]: The model we've used for all this with the law/policy data is to have each one have it's own table, and to define join tables for each metadata-type that can have multiple values associated with a document. Metadata-types with strictly one value have up to this point been implemented as a FK into the appropriate table directly on the document model.

**Comment by joel-wright** [comment]
Thanks for documenting the current state & thoughts. Seems like a reasonable task to pick this up again in the not-too-distant future...

**Comment by joel-wright** [comment]
Closing after moving repos.

**Pull request closed without merging**